rust/private/rustc.bzl

```diff
@@ -320,7 +320,8 @@ def construct_arguments(
         out_dir,
         build_env_files,
         build_flags_files,
-        maker_path = None,
+        clippy_output = None,
```

Comment:
Yes, I would like that too, but the problem is that generating the clippy output requires changing the args to both process_wrapper and rustc. The fact that process wrapper is used to invoke rustc means that the order of the args matters. We can't just append the clippy args in clippy.bzl

There might be a way to do it, but it's not obvious to me despite staring at this code for awhile.

Reply:
Given that ordering doesn't seem to matter for the process wrapper, perhaps `construct_arguments` could take an optional [Args](https://docs.bazel.build/versions/main/skylark/lib/Args.html) object in place of `maker_path`/`clippy_output` and instead of always creating a new set of args, just append to the optional one?

Reply:
I'll give that a try and see what I can come up with.

Reply:
OK...there are 2 things going on here.

- I can move --touch-file / --stdout-file to clippy.bzl and get rid of clippy_output/maker_path (typo for marker_path). No problem. I don't even think I need to pass in optional args.
- I'm not sure how to handle _error_format, which was added [here](https://github.com/bazelbuild/rules_rust/commit/22d67305d80fbc1f0f3e959b3d520e908cf7f9aa). Since it's is a global flag that's an attribute for all of the rust_* rules, I can't move it out of construct_arguments(), so I don't see how to avoid calling the function twice.

I understand, in principle, that _error_format should be useful for IDE/tooling integrations, as the PR indicates. But I don't know how that works in practice, even though I have precisely the same goals. When I started on this work, it seemed like it should be simple:

- Use rust_clippy_aspect
- Set the error format to JSON.

But:

- I'm required to use a specific name for the output group, not "clippy_checks".
- I need the JSON to be written to an output file so I can convert it.
- I don't want warnings to be treated as failures.

I don't know who is using _error_format and rust_clippy_aspect, nor how, so it's hard to refactor in an informed way.

Reply:
Why would you need to move `_error_format` out of `construct_arguments`?

Perhaps `construct_arguments` can take `attrs` as an explicit struct where you could always set that for the clippy action?

so most calls to it would look like
```python
args, env = construct_arguments(
    ctx = ctx,
    # ...
```

But now that'd get moved to
```python
args, env = construct_arguments(
    ctx = ctx,
    attrs = ctx.attrs,
    # ...
```

Except in the case of clippy where you'd actually generate a modified version of `attrs` which doesn't set `_error_format`?

Reply:
(just looping https://github.com/bazelbuild/rules_rust/pull/850#issuecomment-885039975 back to this thread)

Reply:
https://gist.github.com/djmarcin/689246d5c01eeb85c84c28824c27f6a9 is how we are using error_format currently. As you can see it's not a super polished experience by any means. The main goal I had was: Allow rust-analyzer to trigger a build, and be able to convert any output into a format that rust-analyzer expects so it can feed that back across the LSP.

The json error format is a good step, but it's just the rustc output. rust-analyzer expects the cargo output format, which is an additional layer around the rustc output. That's why the extra custom script to convert things is in the gist.

The global error_format flag is useful because it's orthogonal to the use of clippy. You can get regular rustc output in JSON format, or you can get clippy output in human readable format, and vice versa. That being said, I think the idea of having a file sink for the machine readable output to make passing it along to tooling easier is a good one.

Reply:
Apologies for the radio silence. Now that construct_arguments has been refactored, I took another crack at this, and will send a new PR.